### DISCUSSION TITLE
Canvas - draw in local coordinates

### DISCUSSION THREAD

**Discussion by 3elDU** [emilk/egui#3036] [Q&A] [answered]
I have a canvas, but drawing uses global coordinates, not local ones, thus the drawing is relative to the top-left corner of the window, not the canvas:
![image](https://github.com/emilk/egui/assets/63924137/9589cd83-0890-4e73-b692-0de04509406b)
```rust
egui::CentralPanel::default().show(ctx, |ui| {
    ui.label("IT WORKS!!!!!!");
    if ui.button("Do something").clicked() {
        println!("clicked");
    };
    egui::Frame::canvas(ui.style()).show(ui, |ui| {
        ui.painter().rect_filled(
            egui::Rect::from_x_y_ranges(10.0..=50.0, 10.0..=50.0),
            egui::Rounding::none(),
            egui::Color32::from_rgb(70, 120, 200),
        );
    });
});
```
My question is, how to draw in local canvas coordinates, so that (0, 0) would be top-left corner of the canvas?

**Comment by JumpyLionnn** [comment, marked as answer]
You can add the canvas position to every shape you want to draw on it.
you can get the canvas position like this:
```rust
let position = ui.available_rect_before_wrap().min;
```
And apply it to the rect you have there like this: 
```rust
egui::Rect::from_x_y_ranges(10.0..=50.0, 10.0..=50.0).translate(position.to_vec2())
```

I hope i helped :)

**Reply by 3elDU** [reply]
Thanks!
